### ISSUE TITLE
SDK & ASP.Net images contain files & directories with insecure permissions

### ISSUE THREAD

**Issue by paulgear** [dotnet/dotnet-docker#1454]
## Steps to reproduce the issue
1. docker run -ti mcr.microsoft.com/dotnet/core/sdk:2.1-bionic
2. find / -xdev -type d \( -perm -0002 -a ! -perm -1000 \) -ls  # <-- find world-writable directories without the sticky bit set
3. find / -xdev -type f -perm -o+w -ls  # <-- find world-writable files
4. find / -xdev \( -nouser -o -nogroup \) -ls  # <-- find files without a registered user or group

## Expected behavior
Should be no output

## Actual behavior
Here is a non-exhaustive list of dangerous files/directories:
https://pastebin.ubuntu.com/p/P7Kgq6WTf3/

Some of the files in question are .Net DLLs, which I suspect could be used to gain a foothold on a container running externally-facing services and then used to pivot to gain greater access.

## Additional information (e.g. issue happens only occasionally)

This affects -aspnet and -sdk variants of the Docker images, but didn't affect the -runtime versions when I tested.

## Output of `docker version`

Not terribly relevant, but here you go:
```
Client: Docker Engine - Community
 Version:           19.03.4
 API version:       1.40
 Go version:        go1.12.10
 Git commit:        9013bf583a
 Built:             Fri Oct 18 15:52:22 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.4
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.10
  Git commit:       9013bf583a
  Built:            Fri Oct 18 15:50:54 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.5
  GitCommit:        bb71b10fd8f58240ca47fbb579b9d1028eea7c84
 runc:
  Version:          1.0.0-rc6+dev
  GitCommit:        2b18fe1d885ee5083ef9f0838fee39b62d653e30
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

## Output of `docker info`

```
Client:
 Debug Mode: false

Server:
 Containers: 356
  Running: 1
  Paused: 0
  Stopped: 355
 Images: 1413
 Server Version: 19.03.4
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: bb71b10fd8f58240ca47fbb579b9d1028eea7c84
 runc version: 2b18fe1d885ee5083ef9f0838fee39b62d653e30
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 3.10.0-1062.4.1.el7.x86_64
 Operating System: Red Hat Enterprise Linux
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 15.28GiB
 Name: REDACTED
 ID: REDACTED
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by MichaelSimons** [comment]
@paulgear - Thanks for logging this issue.

Here is a summary of the issues:

1. 2.1 & 2.2 SDK - world-writable files - numerous DotnetTools + AppHostTemplate/apphost
1. 2.2 aspnet - world-writable files - Microsoft.AspNetCore.All
1. 3.0 & 3.1 SDK - sticky bit - /tmp/NuGetScratch 

The 2.2 and 2.1 issues have been fixed by the product teams in 3.0 onward and were not back ported.  The 3.* NuGetScratch issue is a direct result of installing PowerShell in the SDK image and is something that should get cleaned up in 3.1.

**Comment by MichaelSimons** [comment]
I forgot to mention this can easily be cleaned up via calling `dotnet nuget locals all --clear` in the `# Install PowerShell global tool` section.

**Comment by MichaelSimons** [comment]
I missed the last one, several files owned by unregistered users/groups.  This should also get fixed in 3.1.  I am concerned of the potential impact of back porting this change.

**Comment by paulgear** [comment]
Hi @MichaelSimons - thanks for looking into this.  Because we're failing our build pipelines when problems with the image are detected, we've manually fixed this in our images (by making the files non-writable, or by assigning the unowned files to root), and we haven't noticed any problems with our workloads as yet.  At presently we're only using `mcr.microsoft.com/dotnet/core/aspnet:2.2-bionic` & `mcr.microsoft.com/dotnet/core/sdk:2.2-bionic` (with plans to go to 3.x soon), so this may not apply to every case, but potential impact of backporting seems low.

**Comment by MichaelSimons** [comment]
Thanks for the response @paulgear.  We have taken a pretty strict stance that we will avoid making breaking changes to released versions of the .NET Core Docker images.  The changes to fix this issue would be considered a breaking change and could break some scenarios.  If users want/need to they can very easily workaround this issue just as you have done.  Based on this, we are only planning to fix the issue within the 3.1 images.

**Comment by mthalman** [comment]
Merged into master for the 3.1 release with https://github.com/dotnet/dotnet-docker/pull/1509.
